### DISCUSSION TITLE
Supported CrossBridge Fork

### DISCUSSION THREAD

**Discussion by Fancy2209** [airsdk/Adobe-Runtime-Support#2843] [Ideas]
[CrossBridge](https://GitHub.com/crossbridge-community/crossbridge)  allowed C/C++ code to be compiled into SWC and SWF, and I believe it's a handy tool even with ANE existing, has HARMAN considered doing a maintained fork of it?

**Comment by marchbold** [comment]
Definitely would be nice to have a maintained version of crossbridge.

**Comment by pol2095** [comment]
Great idea, to convert [lunasvg](https://github.com/airsdk/Adobe-Runtime-Support/discussions/2779#discussioncomment-7067775), I needed to use [adobe-flash/crossbridge branches futures](https://github.com/adobe-flash/crossbridge/tree/futures), support of C++11 but it was never finalized and work only on macOS.

**Reply by marchbold** [reply]
Would be great to put together some documentation on using it and add it to the airsdk.dev site too.

**Reply by pol2095** [reply]
The documentation is in [crossbridge/docs](https://github.com/crossbridge-community/crossbridge/tree/master/docs/) or [here](http://pol2095.free.fr/crossbridge/)
and examples are in [crossbridge/samples](https://github.com/crossbridge-community/crossbridge/tree/futures/samples)

**Reply by Fancy2209** [reply]
I also archived their site at https://fancy2209.github.io/crossbridge.io/

**Comment by Fancy2209** [comment]
Having a maintained CrossBridge version imo could be very rewarding. Allowing more languages in the runtime would be neat, and I think it'd make sense to get C/C++ working again before the proposed TypeScript port
I understand Harman having difficulties about doing this due to the difficulty of contributing to FOSS projects though

**Comment by Matvii-Jarosh** [comment]
I still use CrossBridge to create libraries for SWF files, because this way I am sure that it will work on all devices

**Reply by Fancy2209** [reply]
Yeah, as long as it doesn't need to talk to a native app like steam, no need to have it be a FRE/ANE

**Comment by Fancy2209** [comment]
@ajwfrost has this been considered by HARMAN? (wipro? Teleca? not sure how to address the company since it was acquired) 
FlasCC/CrossBridge was known to produce more optimized code than mxmlc and this has it's benefits (getting C/C++ apps running everywhere without a bunch of ifdefs thanks to AIR, getting libraries working without ANE to not need to manually support all AIR platforms)

**Reply by ajwfrost** [reply]
Teleca is a name from the past :-)  
We had taken a copy of the repository but we'd not looked at actively maintaining it. But I agree it's quite useful for C/C++ based utilities and libraries that don't have particular dependencies on OS-specific interactions.
Whilst I would love for us to get involved in it, I think currently we have plenty of other things that also need attention... Crossbridge is something that (afaik) is fully open source so could be maintained by other folk if it was needed. We could perhaps try to make it easier to install/configure/use though..? let me discuss a bit further with the team on this.

thanks

**Reply by Fancy2209** [reply]
> Teleca is a name from the past :-) We had taken a copy of the repository but we'd not looked at actively maintaining it. But I agree it's quite useful for C/C++ based utilities and libraries that don't have particular dependencies on OS-specific interactions. Whilst I would love for us to get involved in it, I think currently we have plenty of other things that also need attention... Crossbridge is something that (afaik) is fully open source so could be maintained by other folk if it was needed. We could perhaps try to make it easier to install/configure/use though..? let me discuss a bit further with the team on this.
> 
> thanks

The main problem is how bitrotted it is
- Very flaky and big build system, 
- LLVM-GCC 4 and LLVM 2 aren't really new, the LLVM3 and Clang branch is hard to compile and LLVM changed wayy too much to port the changes to current LLVM what I tried, uses cygwin instead of msys2 and only works on a 32 bit version of cygwin (who are no longer maintained and have no official repos up), and it isn't possible to build on Linux, even the crossbridge community fork doesn't seem to work nowdays.
There was a short lived fork, https://github.com/crossbridge-community/crossbridge/, but with how bit rotted it is and the amount of work it'd take to get modernised is VERY large

**Reply by marchbold** [reply]
Must say it's been a while since I looked at it but my experiences mirror yours, the build system is really picky and lots of unmaintained information in there.

**Reply by pol2095** [reply]
The best version : [adobe-flash/crossbridge branches futures](https://github.com/adobe-flash/crossbridge/tree/futures), support of C++11 but it was never finalized and work only on macOS.
I use it on Mac Intel, I never tried it on a Mac ARM.

**Reply by pol2095** [reply]
I tried [https://sourceforge.net/projects/crossbridge/files/](https://sourceforge.net/projects/crossbridge/files/) > Crossbridge_1.1.0.devbuild.dmg on a mac ARM and it work correctly.

just needed `sudo xattr -d -r com.apple.quarantine /Users/user/crossbridge`

**Comment by pol2095** [comment]
Regarding the optimization of native performance and the inherent limitations of the AVM2: rather than updating CrossBridge (which remains a translation layer), have you evaluated the feasibility of reviving the LLVM (AOT) approach that Adobe had initiated with 'Project Next' ?
While the exact extent of Adobe's progress on this front before the project's cancellation remains unclear to the public, I am aware that compiling dynamic AS3 via LLVM is a massive technical challenge, potentially unfeasible without introducing strict typing and Value Types (Structs). However, is the R&D effort required to natively generate direct machine code — similar to what Flutter does with Dart — definitively ruled out of the AIR roadmap, or is it an architectural evolution you are considering ? thanks

**Reply by ajwfrost** [reply]
Hi

I can dig out the old documentation that we'd got from Adobe on the "Next" runtimes, but we never got any of the source code for this...

The idea of using the "ABC" format within a SWF was to ensure they were fully cross-platform compatible. Internally, this gets compiled down into a low-level intermediate representation (i.e. a similar kind of process to LLVM) before that's then JIT-assembed into machine code. The intermediate representation would be different depending on what platform the code is running on, which is again similar to LLVM where you have to give it the details of the CPU/architecture/OS in order for it to use the appropriate/supported set of instructions.

So currently we have a route within the runtime for taking machine-agnostic ABC code and turning it into machine code. There's not a massive overhead for this but the generated machine code is not the most efficient - and crucially, is still running a dynamically typed language which means there are a lot of calls out from the generated code into utility functions so that it can actually resolve the scope of a variable that is referenced in the code.

Ahead-of-time compilation still has this limitation about the code trying to understand what you want to call when you use a function such as "trace()" so the scope chain and lookup/resolution etc is still just dynamically happening at runtime. Which is very different from a C++ type approach where resolution is more structural based on function pointers/tables and the like.

We'd been looking at AOT compilation for Android - in fact, we have the toolchain set up and working, the code is built, but the bootstrapping is not working for some reason. We'd lose the benefit of having one bit of code (within the SWF) that works on all the supported platforms, but given the removal of the 'shared runtime' approach many years ago, that benefit isn't so key now I think. But this approach isn't so compatible with the ".air" installation approach that some people still use on desktops.

Changing any of these kinds of things is actually a significant effort which is a bit of a limitation. But I would definitely be keen to see whether we can update some of the technologies used within AIR to make it somewhat more performant and help expand its reach..

It's certainly a good conversation to have!

**Reply by ajwfrost** [reply]
FYI, the AS4 information that we have is pretty much the same as what's available publicly:
https://github.com/adobe-research/ActionScript4/tree/master

In terms of AS3 to LLVM, this is what happens within the AOT compiler but using a very old version of LLVM. And when we build the old version on a modern computer, it no longer works i.e. identical LLVM IR is created, but whilst the original LLVM binaries that Adobe created will process this properly, newly built LLVM binaries (from the same codebase) throw an error with a function not being identified as a constant (which I suspect is related to changes within C++/STL .. we're seeing if we can get this working somehow as it's kind of needed now for the updates to compile-abc that we need to do...)

All this stuff is also open-source anyway so getting the avmplus code and AOT compiler to build with modern LLVM might be an interesting challenge for someone with a lot of time on their hands... (at least, it may be slightly different from what we have, but there are files such as [AOTCompiler.cpp](https://github.com/adobe-flash/avmplus/blob/master/aot/AOTCompiler.cpp)...)

**Reply by pol2095** [reply]
Regarding the ActionScript 'Next' (AS4) project, it is my understanding that while Adobe shared their research and specifications regarding the language grammar, the actual compilation infrastructure—specifically the LLVM-based backend pipeline designed to translate AS Next into C++ for AOT compilation—remained proprietary and was never handed over to the community. In fact, we can still see the 'fossils' of this abandoned project in the ecosystem today, notably through the ASC 2.0 (Falcon) frontend compiler and the low-level AS3 Intrinsics, but the core engine bridging them to native AOT is missing.

**Reply by ajwfrost** [reply]
In case it's of interest... you can turn ActionScript byte code into LLVM intermediate representation, using the `compile-abc-64` tool that's in the AIR SDK.

* Adding an argument `-filetype=bitcode` would output the .bc file rather than the default of creating an object file.
* Adding `-save-ll` should output a human-readable .ll file (or `-save-raw-ll` is pre-optimisation). You can use `-filetype=null` to make this the only output

We've just been using that to try to work out what was causing the problems in the recent AIR SDK when running the Apple Silicon version of compile-abc-64...
